[Moviepilot] Add extractor

panatexxa opened the pull request:
---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This add support for Trailers on Moviepilot. The whole topic is a little bit complicated:
- Trailers aren't self hosted but hosted at Dailymotion (since 2 years or so).
- Most movies link to a bunch of trailers, for example see [Interstellar](https://www.moviepilot.de/movies/interstellar-2/trailer).
- On the main movie site exactly one "main" trailer is played.

My extractor extracts this "main" trailer as playlist and forwards it to the Dailymotion extractor.
I'm not really sure, if this is a good method. Especially the test cases aren't really expressive because of that.

Moviepilot sometimes changes it's source code. Currently, the video ID of the "main" trailer is only mentioned in a script comment when retrieving the plain website, so I take this. I'm not sure how stable this is but every other method seems to need JS.

panatexxa commented:
Hmm, actually the /trailer webpage seems to be more stable in providing the correct video ID. I will update my extractor.

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,67 @@
+
+
+class MoviepilotIE(InfoExtractor):
+    """Downloader for the trailer for a movie."""
```
```suggestion
```
Set IE_NAME and IE_DESC if needed instead

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,67 @@
+            r'<meta property="og:title" content="(.*)"', webpage, 'title'))
+        embed_url = self._search_regex(r'meta itemprop="embedURL" content="(.*?)"', webpage, 'embed_url')
+
+        return self.playlist_result([self.url_result(embed_url)],
+                                    playlist_id=video_id,
+                                    playlist_title=title)
```
why playlist? You should just return the URL

panatexxa replied:
This was my workaround to trigger the next extractor. This is also the reason why I have no further metadata: Because they are meaningless in playlists.

I have found about `_type: 'url_transparent'` which seems much more suited and updated the extractor with some metadata.

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,67 @@
+
+        webpage = self._download_webpage(url, video_id)
+
+        title = unescapeHTML(self._search_regex(
+            r'<meta property="og:title" content="(.*)"', webpage, 'title'))
```
_og_search_title

pukkandan reviewed (changes requested):
It is better to add `embedURL` extraction into `GenericIE` than creating an extractor, especially since you are not actually extracting any metadata.

pukkandan commented:
Please avoid force-pushing in the middle of a review

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+        },
+    }]
+
+    def _get_property(self, name, webpage, fatal=True):
+        return self._search_regex(f'meta itemprop="{name}" content="(.*?)"', webpage, name, fatal=fatal)
```
this already exists as `self._html_search_meta`

panatexxa replied:
Unfortunately, `self._html_search_meta` is too broad, at least for the `description` field.

For example, in the Interstellar wegpage the following two entities are present:
```
<meta name="description" content="Interstellar Trailer &amp; Teaser, Interviews, Clips und mehr Videos auf Deutsch und im Original. Schaue dir alle 35 Videos jetzt an!"> 
```
```
<meta itemprop="description" content="16 von 35 (Englisch, 2014), Originaltitle: Interstellar, Hard SF, Science Fiction-Film, Abenteuerfilm, von Christopher Nolan, mit Matthew McConaughey, Anne Hathaway" />
```

My function matches the (correct) second one, while `self._html_search_meta` matches also the first one.

pukkandan replied:
then use this only for description (no need to make a function) and add a comment like `_html_search_meta is not used since we don't want name=description to match`

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+
+
+class MoviepilotIE(InfoExtractor):
+    _IE_NAME = "moviepilot"
```
use single quotes everywhere. Make sure you have read the contributing guidelines

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+class MoviepilotIE(InfoExtractor):
+    _IE_NAME = "moviepilot"
+    _IE_DESC = "Moviepilot trailer"
+    _VALID_URL = r'https?://(?:www\.)?moviepilot\.de/movies/(?P<id>[^/]+)/?.*'
```
```suggestion
    _VALID_URL = r'https?://(?:www\.)?moviepilot\.de/movies/(?P<id>[^/]+)'
```
it doesnt do anything

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+    _TESTS = [{
+        'url': 'https://www.moviepilot.de/movies/interstellar-2/',
+        'info_dict': {
+            # 'id': 'interstellar-2',
```
id is always returned by the inner extractor. You can return this as `display_id` if you wish

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+                'title': self._og_search_title(webpage),
+                'url': self._get_property("embedURL", webpage),
+                'thumbnail': self._get_property("thumbnailURL", webpage, fatal=False),
+                # The duration is something like "P00H02M20S" but parse_duration cannot parse this
+                # should I write an extra function for that or should parse_duration be able to parse this?
```
You should ask for clarifications on github, not put as comment in code 😐 

To answer this, use regex to either convert this into a parsable format or even directly convert to seconds

panatexxa replied:
I have copied parts of the code of `parse_duration` now. Is it necessary to convert to float here (also taken from `parse_duration`)?

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+
+        webpage = self._download_webpage(url, video_id)
+
+        return {'_type': 'url_transparent',
```
Read "Trailing parentheses" section of code style guide

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+        video_id = self._match_id(url)
+
+        # normalize url
+        url = f"https://www.moviepilot.de/movies/{video_id}/trailer"
+
+        webpage = self._download_webpage(url, video_id)
```
```suggestion
        webpage = self._download_webpage(f'https://www.moviepilot.de/movies/{video_id}/trailer', video_id)
```

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+
+        return {'_type': 'url_transparent',
+                'ie_key': 'Dailymotion',
+                'id': video_id,
```
don't return wrong id

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+        webpage = self._download_webpage(url, video_id)
+
+        return {'_type': 'url_transparent',
+                'ie_key': 'Dailymotion',
```
```suggestion
                'ie_key': DailymotionIE.ie_key(),
```

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+                # should I write an extra function for that or should parse_duration be able to parse this?
+                # 'duration': parse_duration(self._get_property("duration", webpage, fatal=False))
+                'description': unescapeHTML(self._get_property('description', webpage, fatal=False)),
+                'timestamp': parse_iso8601(self._get_property('uploadDate', webpage, fatal=False),
+                                           delimiter=" "),
```
```suggestion
                'timestamp': parse_iso8601(self._get_property('uploadDate', webpage, fatal=False), delimiter=' '),
```

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+                'description': unescapeHTML(self._get_property('description', webpage, fatal=False)),
+                'timestamp': parse_iso8601(self._get_property('uploadDate', webpage, fatal=False),
+                                           delimiter=" "),
+                'height': int_or_none(self._get_property('height', webpage, fatal=False)),
+                'width': int_or_none(self._get_property('width', webpage, fatal=False))}
```
since the video has multiple formats, this is wrong

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,121 @@
+        return {'_type': 'url_transparent',
+                'ie_key': 'Dailymotion',
+                'id': video_id,
+                # Currently, the movie title is extracted, e.g. "Interstellar" or "Müllers Büro"
+                # The actual trailer title is: "Interstellar - Trailer 2 (Deutsch) HD" or
+                # "MÃ¼llers BÃ¼ro - Trailer (Deutsch)", umlauts are really broken here.
+                # I prefer to keep the actual movie title or is there an extra metadata field for that?
+                'title': self._og_search_title(webpage),
```
this is fine

pukkandan replied:
fyi: if you want to extract both titles, one can be `alt_title`

pukkandan reviewed (changes requested):
also, none of the comments look useful. Remove them

panatexxa commented:
> Please avoid force-pushing in the middle of a review

Sorry, I thought that Github will track the changes between the force pushes. Will I squash everything at the end then?

pukkandan commented:
> Sorry, I thought that Github will track the changes between the force pushes.

It does track the changes, but it still makes reviewing harder

> Will I squash everything at the end then?

I will squash and merge

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -0,0 +1,119 @@
+
+        webpage = self._download_webpage(f'https://www.moviepilot.de/movies/{video_id}/trailer', video_id)
+
+        # oriented on parse_duration()
+        d = re.match(r'P(?P<hours>[0-9]+)H(?P<mins>[0-9]{2})M(?P<secs>[0-9]{2})S',
+                     self._get_property("duration", webpage, fatal=False))
+        if d:
+            hours, mins, secs = d.groups()
+            duration = float(hours) * 60 * 60 + float(mins) * 60 + float(secs)
+        else:
+            duration = None
```
```suggestion
        duration = try_get(
            re.match(r'P(?P<hours>\d+)H(?P<mins>\d+)M(?P<secs>\d+)S',
                     self._html_search_meta('duration', webpage, fatal=False) or ''),
            lambda mobj: sum(float(x) * y for x, y in zip(mobj.groups(), (3600, 60, 1))))
```
(can be split into 2 expressions if u think it's too complicated)

panatexxa replied:
Complexity is fine so.

What do you think of?
```diff
- lambda mobj: sum(float(x) * y for x, y in zip(mobj.groups(), (3600, 60, 1))))
+ lambda mobj: float(sum(x * y for x, y in zip(mobj.groups(), (3600, 60, 1))))
```
Is the float conversion necessary at all?

pukkandan replied:
it won't work since values returned by regex are strings

panatexxa replied:
Ok.

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -23,7 +24,7 @@ class MoviepilotIE(InfoExtractor):
             'ext': 'mp4',
             'title': 'Interstellar',
-            'thumbnail': 'https://s1.dmcdn.net/v/SaXev1VvzitVZMFsR/x720',
+            'thumbnail': r're:https://.*\.dmcdn\.net/v/SaXev1VvzitVZMFsR/x720',
```
```suggestion
            'thumbnail': r're:https://\w+\.dmcdn\.net/v/SaXev1VvzitVZMFsR/x720',
```
same elsewhere

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -89,31 +90,27 @@ class MoviepilotIE(InfoExtractor):
+                     self._html_search_meta('duration', webpage, fatal=False) or ''),
+            lambda mobj: sum(float(x) * y for x, y in zip(mobj.groups(), (3600, 60, 1))))
+        # _html_search_meta is not used since we don't want name=description to match
+        description = unescapeHTML(
+            self._search_regex(f'meta itemprop="description" content="(.*?)"', webpage, 'description', fatal=False))
```
```suggestion
        description = _html_search_regex(
            '<meta[^>]+itemprop="description"[^>]+content="([^>"])"', webpage, 'description', fatal=False)
```

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -89,31 +90,27 @@ class MoviepilotIE(InfoExtractor):
-            'thumbnail': self._get_property('thumbnailURL', webpage, fatal=False),
-            'description': unescapeHTML(self._get_property('description', webpage, fatal=False)),
+            'url': self._html_search_meta('embedURL', webpage),
+            'thumbnail': self._html_search_meta('thumbnailURL', webpage, fatal=False),
```
```suggestion
            'thumbnail': self._html_search_meta('thumbnailURL', webpage),
```
that is the default!

pukkandan commented on yt_dlp/extractor/moviepilot.py:
```diff
@@ -89,31 +90,27 @@ class MoviepilotIE(InfoExtractor):
+            'description': description,
             'duration': duration,
-            'timestamp': parse_iso8601(self._get_property('uploadDate', webpage, fatal=False), delimiter=' ')
+            'timestamp': parse_iso8601(self._html_search_meta('uploadDate', webpage, fatal=False), delimiter=' ')
```
```suggestion
            'timestamp': parse_iso8601(self._html_search_meta('uploadDate', webpage), delimiter=' ')
```

panatexxa commented:
This should address all your comments. Just curious: Up to now, I have resolved the conversation, when I thought that I have addressed your comment. Is this the normal way or do you resolve the conversation, when you think that I have correctly fixed the code?

pukkandan commented:
Doesn't really matter as long as the comments get addressed

panatexxa commented:
Thank you very much for the detailed review and maintaining in general!
